Network graph

pulkit10251 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
One can now visualize Network graphs, nodes, and edges using the viz.graph() function. The user will have the option to visualize the graph as an undirected normal graph, directed normal graph, weighted graph, directed-weighted graph. They will have control to show edge and vertex labels.
## Description
<!--- Describe your changes in detail -->
In this pull request, I have made changes in the library at 8 places. The changes are : 
**1.  server.py** :  
Added code in def window() function to handle the opts parameters. Also added downloadable scripts for these libraries: 
1. d3.v3.min.js  (for implementing d3 based network visualization)
2. d3-selection-multi.v1.js (to set multiple CSS properties for d3 graph)
3. saveSvgAsPng.js (to be able to download SVG image in PNG format)

**2. Network.css**:
Added CSS for NetworkPane.js.

**3. Index.html**: 
Added scripts and css for the above-downloaded libraries and newly created css file.   

**4. NetworkPane.js**:
Created a code to visualize graphs in javascript. It was done using the d3 library.

**5.  init.py**
Created a function def graph() to get data from the user and validate the user input using the library "cerberus" and send the validated data to the server.

**6.  init.pyi**:
Added definition for the newly created graph.

**7. setup.py**:
Added the "cerberus" library in the requirements.

**8. demo.py**:  
Added demo code for graph().

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
Till this pull request, it was not possible to visualize graphs in visdom, though it was possible to display non-interactive image using networkx library, there is no way to display network-graph, such that it remains interactive with the user. Thus this pull request is a feature addition to visdom.
<!--- If it fixes an open issue, please link to the issue here. -->
Yes, it fixes this issue. this pull request is made in response to issue #805. 
## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, experiments you ran to see how -->
<!--- your change affects existing areas of the code and their behaviors, etc. -->
<!--- One method of testing is to run the `demo.py` script from the examples -->
<!--- both on your branch and a clean branch and ensure that none of the functionality -->
<!--- appears different. Be sure to install from source when testing. -->
 I have run demo.py to test the functionality and it is working fine. 
## Screenshots (if appropriate):
![plot (5)](https://user-images.githubusercontent.com/54267595/131233394-06940be6-8179-45a6-b94b-60c3a5922afa.png)

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] For JavaScript changes, I have re-generated the minified JavaScript code.

JackUrb commented on py/visdom/__init__.pyi:
```diff
@@ -137,3 +137,10 @@ class Visdom:
+        win: _OptStr = ..., 
+        env: _OptStr = ..., 
+        opts: _OptOps = ...
+    ) -> _SendReturn: ...
```
Nit: retain newlines at the ends of files.

JackUrb commented on py/visdom/static/css/network.css:
```diff
@@ -0,0 +1,12 @@
+  .link {
+    stroke: #999;
+    stroke-opacity: 0.6;
+  }
```
Nit: Newline

JackUrb commented on py/visdom/__init__.py:
```diff
@@ -2312,3 +2318,115 @@ def dual_axis_lines(self, X=None, Y1=None, Y2=None, opts=None, win=None, env=Non
+            "links": links
+        }
+
+        jsonstring = json.dumps(GData)
+        GData = json.loads(jsonstring)
```
Any reason for the dump and reload here? I tend to prefer comments for the rationale on things that are 'no-ops' outside of externalities.

dizcza replied:
Right. I think this part was done for debugging and therefore could and should be removed.
Or did it have any other purpose?

pulkit10251 replied:
Nicely pointed out @JackUrb  and @dizcza. It was done for debugging purpose only. I will make the changes. 
Thanks

JackUrb reviewed (approved):
This feature is awesome, thanks for the contribution @pulkit10251. I've wanted visdom to have some kind of integrated graph functionality for a long while, but just never was able to work on it. There are many who I am sure will use this, especially seeing it in the demo you added to our `demo.py`. 

The only thing I'd like to see more of in here is your documentation or thought process, especially with respect to the `NetworkPane` implementation and to a lesser degree to the logic parsing in `graph` (as you already have a fairly sufficient amount there). Visdom relies heavily on community wants and contributions, and keeping our codebase self-documented is key to ensuring that your functionality is maintainable in the future.

This is approved from my POV, though I'll leave a moment for @dizcza to add any comments considering he had done a deeper exploration of this in the past.

dizcza commented on js/NetworkPane.js:
```diff
@@ -0,0 +1,221 @@
+  };
+
+  CreateNetwork = graph => {
+    var width = 500,
+      height = 500;
```
width and height attributes should be extracted from the `opts` dict, shouldn't they?

pulkit10251 replied:
Hi @dizcza , Actually there is a error in it which i forgot to mention, Actually the problem is networkPane is getting props.height and props.width as Null. So to remove it temporarily I have given a 500 value to them. It would be nice if someone can check why it is happening. 
Thanks

pulkit10251 replied:
I am just looking at the code, and I have noticed that height and width are undefined in PlotPane also. I also explored the possibility of using this.props.w and this.props.h but they are very small (24, 30) in my case. so is there any conversion that I am missing?

dizcza replied:
Sadly, I cannot comment on this issue. Perhaps, @JackUrb could clarify this.

JackUrb replied:
Indeed `w` and `h` are scaled, in this case to be able to adapt to the `ReactGridLayout` cells rather than px sizes. These are computed for the layout of the current added pane here: https://github.com/fossasia/visdom/blob/master/js/main.js#L371-L373

dizcza commented on js/NetworkPane.js:
```diff
@@ -0,0 +1,221 @@
+      });
+
+    if (this.props.showEdgeLabels) {
+      if (this.props.labels === 'custom') {
+        edgelabels
+          .append('textPath')
+          .attr('xlink:href', (d, i) => '#edgepath' + i)
+          .style('text-anchor', 'middle')
+          .style('pointer-events', 'none')
+          .attr('startOffset', '50%')
+          .text(d => d.label);
+      } else {
+        edgelabels
+          .append('textPath')
+          .attr('xlink:href', (d, i) => '#edgepath' + i)
+          .style('text-anchor', 'middle')
+          .style('pointer-events', 'none')
+          .attr('startOffset', '50%')
+          .text(d => d.label);
```
the branches code is identical: the `if` does not do anything

pulkit10251 replied:
Thanks for pointing this out! I will make the changes.

dizcza commented on js/NetworkPane.js:
```diff
@@ -0,0 +1,221 @@
+        .attr('height', height);
+    }
+
+    if (this.props.Gtype == 'dgph' || this.props.Gtype == 'dwgph') {
```
`Gtype` string can be easily substituted by the boolean `directed` key in `opts` dict.
The execution pipeline enters this block of code iff the graph is directed.

pulkit10251 replied:
Thanks for this suggestion @dizcza, I think it would be best to have this in opts.

dizcza commented:
## Major comments
* You're doing lots of converting from your `graphDict` structure to essentially a dict of `{"nodes": nodes, "edges": edges}`.  I suggest making these two arguments - `nodes` and `edges` - required. By doing so you avoid lots of computations that can be significant for large graphs.
* Let's reap what the best Python graph package - NetworkX - already has. If you're a user of NetworkX, and I don't know any other better Python package for working with graphs, then our API matches perfectly with NetworkX. As easy as
  ```python
   import networkx as nx
   G = nx.Graph()
   G.add_edge("a", "b", weight=0.6)
   G.add_edge("a", "c", weight=0.2)
   viz.graph(G.nodes, G.edges)
   viz.graph(G.nodes, G.edges(data=True))  # weighted graph
   ```
  Please take a look at NetworkX if you're not familiar with it.
* I liked `pyvis` package because it displays the label of a vertex on hover (while the mouse pointer reaches its location in an HTML page). Same for edges. Is it possible to implement this? If not, then remove `"hover"` from the list of available options and make `"center"` (or you're calling this `"middle"`) by default for `showVertexLabels` and `showEdgeLabels`.

## Minor comments
* I really liked the validation scheme, I did it in the past and it does help. But I'm not sure if it fits the design when the design is changing rapidly as it is currently experiencing (from `graphDict` to `nodes` and `edges`). Feel free to rewrite the validation logic again. For now, I removed this part though I liked it :+1: 

To facilitate the ease of merging your ideas with mine, I opened [a pull request](https://github.com/pulkit10251/visdom/pull/1) from my local branch to your `network-graph` local branch. **If you just merge the changes, it will break the functional.** Because I put the TODOs in NetworkPane.js that I don't know how to implement.

pulkit10251 commented:
> ## Major comments
> 1. You're doing lots of converting from your `graphDict` structure to essentially a dict of `{"nodes": nodes, "edges": edges}`.  I suggest making these two arguments - `nodes` and `edges` - required. By doing so you avoid lots of computations that can be significant for large graphs.
> 2. Let's reap what the best Python graph package - NetworkX - already has. If you're a user of NetworkX, and I don't know any other better Python package for working with graphs, then our API matches perfectly with NetworkX. As easy as
>    ```python
>    import networkx as nx
>    G = nx.Graph()
>    G.add_edge("a", "b", weight=0.6)
>    G.add_edge("a", "c", weight=0.2)
>    viz.graph(G.nodes, G.edges)
>    viz.graph(G.nodes, G.edges(data=True))  # weighted graph
>    ```
>    
>    
>        
>          
>        
>    
>          
>        
>    
>        
>      
>    Please take a look at NetworkX if you're not familiar with it.
> 3. I like `pyvis` package because it displays the label of a vertex on hover (while the mouse pointer reaches its location in an HTML page). Same for edges. Is it possible to implement this? If not, then remove `"hover"` from the list of available options and make `"center"` (or you're calling this `"middle"`) by default for `showVertexLabels` and `showEdgeLabels`.
>    Here is an illustration of what pyvis can do; that's a high level of interaction with the user: everything is hidden and neat and yet you can extract information (a label or any text you saved in a graph) by hovering the mouse onto edges and vertices.
>    ```python
>    from pyvis.network import Network  # pip install pyvis
>    net = Network(height="960px", width="1280px", directed=True, layout=True)
>    net.add_node(1, label='a', color='red', title='a title')
>    net.add_node(2, label='b', color='green', title='b title')
>    net.add_node(3, label='c', color='green', title='c title')
>    net.add_edge(1, 2, title='a-b edge', color='blue')
>    net.add_edge(2, 3, title='c-b edge', color='blue')
>    net.show(name='/tmp/a.html')
>    ```
>    
>    
>        
>          
>        
>    
>          
>        
>    
>        
>      
>    
>    
>    
>    ![Screenshot from 2021-09-01 09-25-00](https://user-images.githubusercontent.com/7688337/131622854-5bfcd2fd-c084-4f44-8cb8-7f53567b988f.png)
>    ![Screenshot from 2021-09-01 09-25-10](https://user-images.githubusercontent.com/7688337/131622857-2aa950c3-c40c-41f2-8450-19a7c931a560.png)
>    
>    
>    
>    In the first example, I moved the mouse onto the 'c-b' edge; in the second - onto the 'c' node.
> 4. The user can pass edges as simply pairs of `(source, target)` nodes or, if he/she wants to display additional info, the third element can be a dict that contains the weight, the label, and other attributes the user might like to have in an edge. In other words, whether the user wants to display information about the weight (as well as another attribute) of an edge, depends whether he or she passed `edges` as a list of `(source, target)` pairs or it has the third element - a dict of the attributes of an edge. In the example above,
>    ```python
>    >>> G.edges
>    EdgeView([('a', 'b'), ('a', 'c')])
>    >>> G.edges(data=True)
>    EdgeDataView([('a', 'b', {'weight': 0.6}), ('a', 'c', {'weight': 0.2})])
>    >>> G.add_edge("b", "d", weight=1.2, label='mylabel')
>    >>> G.edges(data=True)
>    EdgeDataView([('a', 'b', {'weight': 0.6}), ('a', 'c', {'weight': 0.2}), ('b', 'd', {'weight': 1.2, 'label': 'mylabel'})])
>    ```
>    
>    
>        
>          
>        
>    
>          
>        
>    
>        
>      
>    If there is a third element in `edges`, we check whether it's a dict and if it's a dict (it must be a dict; otherwise raise error) we extract the `label` key, if it has such a key, and use this key to name the edge. Otherwise, if the dict does not have `label`, we don't label this particular edge. **Is it possible to implement such logic in JS**? Any objections are welcome for discussion.
> 
> ## Minor comments
> * I really liked the validation scheme, I did it in the past and it does help. But I'm not sure if it fits the design when the design is changing rapidly as it is currently experiencing (from `graphDict` to `nodes` and `edges`). Feel free to rewrite the validation logic again. For now, I removed this part though I liked it 👍
> * I replaced str `opts.type` by boolean `opts.directed`.
> * I changed what `showVertexLabels` and `showEdgeLabels` are responsible for because
>   
>   1. The vertices of a graph should be always drawn.
>   2. The edge labels can be drawn if the user wants to.
>   3. How to otherwise specify _how_ and _where_ the label of a vertex or an edge must be drawn?
> * I could not understand the need for the global `label` in the `opts`. Therefore, I removed it. Each edge should have a corresponding label if any (it's acceptable to have none). It doesn't make sense to draw the same label over all edges. Well, perhaps I misunderstood its concept.
> 
> ## Opened questions
> 1. The original and mine API doesn't provide any choice for the positions of the nodes that specify where the graph must be drawn. Don't know how to deal with it. Are you currently choosing the nodes' positions randomly?
> 
> To facilitate the ease of merging your ideas with mine, I opened [a pull request](https://github.com/pulkit10251/visdom/pull/1) from my local branch to your `network-graph` local branch. **If you just merge the changes, it will break the functional.** Because I put the TODOs in NetworkPane.js that I don't know how to implement. I'm not asking to blindly follow the pipeline I wrote - it is still only a sketch...



> ## Major comments
> 1. You're doing lots of converting from your `graphDict` structure to essentially a dict of `{"nodes": nodes, "edges": edges}`.  I suggest making these two arguments - `nodes` and `edges` - required. By doing so you avoid lots of computations that can be significant for large graphs.
> 2. Let's reap what the best Python graph package - NetworkX - already has. If you're a user of NetworkX, and I don't know any other better Python package for working with graphs, then our API matches perfectly with NetworkX. As easy as
>    ```python
>    import networkx as nx
>    G = nx.Graph()
>    G.add_edge("a", "b", weight=0.6)
>    G.add_edge("a", "c", weight=0.2)
>    viz.graph(G.nodes, G.edges)
>    viz.graph(G.nodes, G.edges(data=True))  # weighted graph
>    ```
>    
>    
>        
>          
>        
>    
>          
>        
>    
>        
>      
>    Please take a look at NetworkX if you're not familiar with it.
> 3. I like `pyvis` package because it displays the label of a vertex on hover (while the mouse pointer reaches its location in an HTML page). Same for edges. Is it possible to implement this? If not, then remove `"hover"` from the list of available options and make `"center"` (or you're calling this `"middle"`) by default for `showVertexLabels` and `showEdgeLabels`.
>    Here is an illustration of what pyvis can do; that's a high level of interaction with the user: everything is hidden and neat and yet you can extract information (a label or any text you saved in a graph) by hovering the mouse onto edges and vertices.
>    ```python
>    from pyvis.network import Network  # pip install pyvis
>    net = Network(height="960px", width="1280px", directed=True, layout=True)
>    net.add_node(1, label='a', color='red', title='a title')
>    net.add_node(2, label='b', color='green', title='b title')
>    net.add_node(3, label='c', color='green', title='c title')
>    net.add_edge(1, 2, title='a-b edge', color='blue')
>    net.add_edge(2, 3, title='c-b edge', color='blue')
>    net.show(name='/tmp/a.html')
>    ```
>    
>    
>        
>          
>        
>    
>          
>        
>    
>        
>      
>    
>    
>    
>    ![Screenshot from 2021-09-01 09-25-00](https://user-images.githubusercontent.com/7688337/131622854-5bfcd2fd-c084-4f44-8cb8-7f53567b988f.png)
>    ![Screenshot from 2021-09-01 09-25-10](https://user-images.githubusercontent.com/7688337/131622857-2aa950c3-c40c-41f2-8450-19a7c931a560.png)
>    
>    
>    
>    In the first example, I moved the mouse onto the 'c-b' edge; in the second - onto the 'c' node.
> 4. The user can pass edges as simply pairs of `(source, target)` nodes or, if he/she wants to display additional info, the third element can be a dict that contains the weight, the label, and other attributes the user might like to have in an edge. In other words, whether the user wants to display information about the weight (as well as another attribute) of an edge, depends whether he or she passed `edges` as a list of `(source, target)` pairs or it has the third element - a dict of the attributes of an edge. In the example above,
>    ```python
>    >>> G.edges
>    EdgeView([('a', 'b'), ('a', 'c')])
>    >>> G.edges(data=True)
>    EdgeDataView([('a', 'b', {'weight': 0.6}), ('a', 'c', {'weight': 0.2})])
>    >>> G.add_edge("b", "d", weight=1.2, label='mylabel')
>    >>> G.edges(data=True)
>    EdgeDataView([('a', 'b', {'weight': 0.6}), ('a', 'c', {'weight': 0.2}), ('b', 'd', {'weight': 1.2, 'label': 'mylabel'})])
>    ```
>    
>    
>        
>          
>        
>    
>          
>        
>    
>        
>      
>    If there is a third element in `edges`, we check whether it's a dict and if it's a dict (it must be a dict; otherwise raise error) we extract the `label` key, if it has such a key, and use this key to name the edge. Otherwise, if the dict does not have `label`, we don't label this particular edge. **Is it possible to implement such logic in JS**? Any objections are welcome for discussion.
> 
> ## Minor comments
> * I really liked the validation scheme, I did it in the past and it does help. But I'm not sure if it fits the design when the design is changing rapidly as it is currently experiencing (from `graphDict` to `nodes` and `edges`). Feel free to rewrite the validation logic again. For now, I removed this part though I liked it 👍
> * I replaced str `opts.type` by boolean `opts.directed`.
> * I changed what `showVertexLabels` and `showEdgeLabels` are responsible for because
>   
>   1. The vertices of a graph should be always drawn.
>   2. The edge labels can be drawn if the user wants to.
>   3. How to otherwise specify _how_ and _where_ the label of a vertex or an edge must be drawn?
> * I could not understand the need for the global `label` in the `opts`. Therefore, I removed it. Each edge should have a corresponding label if any (it's acceptable to have none). It doesn't make sense to draw the same label over all edges. Well, perhaps I misunderstood its concept.
> 
> ## Opened questions
> 1. The original and mine API doesn't provide any choice for the positions of the nodes that specify where the graph must be drawn. Don't know how to deal with it. Are you currently choosing the nodes' positions randomly?
> 
> To facilitate the ease of merging your ideas with mine, I opened [a pull request](https://github.com/pulkit10251/visdom/pull/1) from my local branch to your `network-graph` local branch. **If you just merge the changes, it will break the functional.** Because I put the TODOs in NetworkPane.js that I don't know how to implement. I'm not asking to blindly follow the pipeline I wrote - it is still only a sketch...

Hi @dizcza . I hope you are doing great. I have started working on it again and have fixed the resize issue and now I will be working on Python code from now on ( which I will complete most probably by 24). Also, I want to point out one thing that hover  edge title for edges is difficult in d3 while its already present on nodes. what I meant with "showEdgeLabels" or "showNodeLabels" is for whether to display nodes label or not. and the "labels" that I earlier used is for the users to put custom labels in Nodes as most nodes are only numerical. so when user set labels as "custom" then it needs to provide custom data for nodes also.
I am working on for a different style for input against the "adjancy matrix"  style that I used earlier.
please let me know if you have some different ideas. 
I have now worked on the clone from your repo. I will be creating a new pull request in the main visdom repo by integrating these changes.

Thanks

dizcza commented:
@pulkit10251 
glad you're making progress on it.
It's hard to say what API would suit best and what interface the users would like most. Seems like we're the only ones who participate in the decisions to take, and it's not enough to have it polished.
I just know that the community would like to have plug-in-place support for networkx-like graph representation (`G.nodes` and `G.edges`) rather than having the adjacency matrix style or any other style they are not got used to.
For any other things, go on as you prefer / as you see them.

pulkit10251 commented:
> @pulkit10251 glad you're making progress on it. It's hard to say what API would suit best and what interface the users would like most. Seems like we're the only ones who participate in the decisions to take, and it's not enough to have it polished. I just know that the community would like to have plug-in-place support for networkx-like graph representation (`G.nodes` and `G.edges`) rather than having the adjacency matrix style or any other style they are not got used to. For any other things, go on as you prefer / as you see them.

yes I am looking for something very similar to :
for Edge  :
(0, 1, {"weight" : 10,  "label" : "A" })
for  Node : 
(0, "Label")

and please let me know if we can include labels again in opts,  as for graph a person may want to display custom "label", while some users may want to use the weights as the label itself. or we should only provide for weights as label is case weight is provided if weights not present then only use custom labels given. or one thing what I can think of is instead of having two fields "weight" and "label", we can only provide one as "label", If a person wants weight then if can put it into the "labels" as only one will be displayed. What's your take on this?

dizcza commented:
What are the "weight" keys for? Unless you use these values to simulate physical graphs with nodes attracting each other with the weights specified (please don't), the weights as very other keys, save for "label", in the edges dict should be ignored.

I don't think placing the weights, if labels are not given, as edge titles is a good idea. An edge can contain lots of other keys, and surely you don't want to show all of them. Therefore, let's stick to either the labels are given / present in the edges OR don't put anything on the edge title.

Of course, you can include the custom labels parameter again. I don't understand though how you handle the case when the user wants to specify the custom edge labels but not the node ones (leave node labels to default) and vice versa. Or two customs at once. Neither do I think that specifying anything as *custom* will bring you to a clean and simple API.

**Disclaimer**

I'm not affiliated with the Visdom project. Therefore, I can only advise rather than state how it must be. And my suggestions can be wrong and biased.
Do as you feel it.

pulkit10251 commented:
Hi @dizcza  and  @JackUrb , I am done with the work. I have created a separate pull request. Please let me know if we can close this pull request and move to the new one. 
Thanks

pulkit10251 commented:
Hi All, this issue has been contained in a new pull request #826. Please route there for further information.
